### PR TITLE
Add new cop `Style/EndlessMethod`.

### PR DESCRIPTION
**Pull request by dvandersluis**
`Style/EndlessMethod` by default enforces that endless methods definitions must be on a single line (`EnforcedStyle: allow`). It can also be configured to disallow endless method definitions (`EnforcedStyle: disallow`), or to allow every endless  method regardless of lines (`EnforcedStyle: allow_always`).

This cop (obviously) only applies if `TargetRubyVersion` is at least 3.0.

There is a style guide issue open for this: rubocop-hq/ruby-style-guide#858.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop-hq/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by dvandersluis** [comment]
The documentation check failure is already fixed in #9276 (I can cherry-pick the fix commit into here as well, or once that PR is merged I'll rebase).

**Comment by bbatsov** [comment]
The cop looks solid overall, but I'm wondering if this should be a standalone cop or some configuration of the `SingleLineMethods` cop. As endless methods are always single-line it makes sense to me to just fold them under the existing cop and let the users decide if they want to stick to classic approach or endless methods via different enforced styles.

**Comment by bbatsov** [comment]
After getting [this feedback](https://github.com/rubocop-hq/ruby-style-guide/pull/858#issuecomment-750179926) from @mame I'm thinking the cop should actually be based on it. We can have two supported styles:

* allow the new syntax in the cases outlined by @mame (default)
* allow it always
* forbid it always

In light of this, you can forget about my suggestion to merge this cop with `SingleLineMethod`. I didn't think of the case when you can write a multi-line endless method.

**Comment by dvandersluis** [comment]
Just to avoid the eventual bikeshed 😅 what are we going to define the allowed default case as? (I probably won't get to this for at least a couple days but I might have time who knows).

**Comment by bbatsov** [comment]
I was thinking by default we can allow only single-line endless methods and refine the definition later to potentially account for definition simplicity and side-effects (e.g. based on configurable list of side-effect producing methods).

**Comment by dvandersluis** [comment]
Totally agree. I’ll make it happen.

**Comment by dvandersluis** [comment]
@bbatsov I've redone this cop based on the above comments. I removed the changes to `SingleLineMethod`,  but I'm going to open a PR for that too to allow single line methods to be rewritten to endless methods.

The styles I ended up with are `allow` (only single line), `allow_always` (any line length),  and `disallow`. I used `allow` to represent the above discussion (since we might change how it applies based on feedback), but if we don't want to change it, `allow_single_line` might be a better name.

**Comment by koic** [comment]
> `allow_single_line` might be a better name.

I like this better name. I think that `allow_single_line` is clearer and good than `allow`.

**Comment by bbatsov** [comment]
Yeah, I agree. Other than this, the cop seems solid to me.

**Comment by dvandersluis** [comment]
@bbatsov @koic updated to `allow_single_line`.

File: config/default.yml [hunk 1]
```diff
@@ -3196,6 +3196,17 @@ Style/EndBlock:
   VersionAdded: '0.9'
   VersionChanged: '0.81'
 
+Style/EndlessMethod:
+  Description: 'Avoid the use of multi-lined endless method definitions.'
+  StyleGuide: '#no-endless-methods'
```
Comment (bbatsov) [inline config/default.yml:3201]: Btw, we'll also have to update the guideline and its anchor. :-)

Comment (dvandersluis) [reply config/default.yml:3201]: Should I change it to `#endless-methods` or remove it?

Comment (bbatsov) [reply config/default.yml:3201]: I'd go with option 1, as I assume we'll update your guideline proposal and merge it.

Comment (dvandersluis) [reply config/default.yml]: ✅ done

**Pull request merged**